src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized list of items.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class Virtualize<TItem> : ComponentBase, IVirtualizeJsCallbacks, IAsyncDisposable
+    {
+        private VirtualizeJsInterop? _jsInterop;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        private int _itemsBefore;
+
+        private int _visibleItemCapacity;
+
+        private int _itemCount;
+
+        private int _loadedItemsStartIndex;
+
+        private IEnumerable<TItem>? _loadedItems;
+
+        private Task _refreshTask = Task.CompletedTask;
+
+        private CancellationTokenSource? _refreshCts;
+
+        private Exception? _refreshException;
+
+        private ItemsProviderDelegate<TItem> _itemsProvider = default!;
+
+        private RenderFragment<TItem>? _itemTemplate;
+
+        private RenderFragment<PlaceholderContext>? _placeholder;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem>? ChildContent { get; set; }
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem>? Item { get; set; }
+
+        /// <summary>
+        /// Gets or sets the template for items that have not yet been loaded in memory.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<PlaceholderContext>? Placeholder { get; set; }
+
+        /// <summary>
+        /// Gets the size of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets or sets the function providing items to the list.
+        /// </summary>
+        [Parameter]
+        public ItemsProviderDelegate<TItem> ItemsProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the fixed item source.
+        /// </summary>
+        [Parameter]
+        public ICollection<TItem> Items { get; set; } = default!;
```

Comment:
These should be nullable since only one of `ItemsProvider` or `Items` is required.